data-prepper-plugins/sns-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/SNSSinkService.java

```diff
@@ -0,0 +1,234 @@
+
+    private final Counter numberOfRecordsSuccessCounter;
+
+    private final Counter numberOfRecordsFailedCounter;
+
+    /**
+     * @param snsSinkConfig sns sink related configuration.
+     * @param bufferFactory factory of buffer.
+     * @param snsClient
+     * @param pluginMetrics metrics.
+     * @param pluginFactory
+     * @param pluginSetting
+     */
+    public SNSSinkService(final SNSSinkConfig snsSinkConfig,
+                          final BufferFactory bufferFactory,
+                          final SnsClient snsClient,
+                          final PluginMetrics pluginMetrics,
+                          final PluginFactory pluginFactory,
+                          final PluginSetting pluginSetting) {
+        this.snsSinkConfig = snsSinkConfig;
+        this.bufferFactory = bufferFactory;
+        this.snsClient = snsClient;
+        this.reentrantLock = new ReentrantLock();
+        this.bufferedEventHandles = new LinkedList<>();
+        this.maxEvents = snsSinkConfig.getThresholdOptions().getEventCount();
+        this.maxBytes = snsSinkConfig.getThresholdOptions().getMaximumSize();
+        this.maxCollectionDuration = snsSinkConfig.getThresholdOptions().getEventCollectTimeOut().getSeconds();
+        this.topicName = snsSinkConfig.getTopicArn();
+        this.maxRetries = snsSinkConfig.getMaxUploadRetries();
+        this.pluginSetting = pluginSetting;
+
+        this.numberOfRecordsSuccessCounter = pluginMetrics.counter(NUMBER_OF_RECORDS_FLUSHED_TO_SNS_SUCCESS);
+        this.numberOfRecordsFailedCounter = pluginMetrics.counter(NUMBER_OF_RECORDS_FLUSHED_TO_SNS_FAILED);
+
+        this.dlqPushHandler = new DlqPushHandler(snsSinkConfig.getDlqFile(), pluginFactory,
+                String.valueOf(snsSinkConfig.getDlqPluginSetting().get(BUCKET)),
+                snsSinkConfig.getDlqStsRoleARN()
+                ,snsSinkConfig.getDlqStsRegion(),
+                String.valueOf(snsSinkConfig.getDlqPluginSetting().get(KEY_PATH)));
+    }
+
+    /**
+     * @param records received records and add into buffer.
+     */
+    void output(Collection<Record<Event>> records) {
+        reentrantLock.lock();
+        if (currentBuffer == null) {
+            currentBuffer = bufferFactory.getBuffer();
+        }
+        try {
+            for (Record<Event> record : records) {
+                final Event event = record.getData();
+                final byte[] encodedBytes = event.toJsonString().getBytes();
+                currentBuffer.writeEvent(encodedBytes);
+                if (event.getEventHandle() != null) {
+                    bufferedEventHandles.add(event.getEventHandle());
+                }
+                if (checkThresholdExceed(currentBuffer, maxEvents, maxBytes, maxCollectionDuration)) {
+                    LOG.debug("Writing {} to SNS with {} events and size of {} bytes.",
+                            currentBuffer.getEventCount(), currentBuffer.getSize());
+                    String errorMsg = "";
```

Comment:
Why is this passed to `retryFlushToSNS`?

Reply:
Thanks @graytaylor0 for your feedback. we are capturing the exception from the retryFlashToSNS and using it for dlq message. we have changed the logic in new commit. please review and let us your feedback.